internal/konnect/license/storage.go

```diff
@@ -0,0 +1,105 @@
+package license
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v3/internal/license"
+)
+
+const (
+	// licenseResourceNamePrefix is the prefix of the secret name storing the konnect license.
+	licenseResourceNamePrefix = "konnect-license-"
+)
+
+// Storer is the interface to store license fetched from Konnect and load license from storage if failed to fetch.
+type Storer interface {
+	Store(context.Context, license.KonnectLicense) error
+	Load(context.Context) (license.KonnectLicense, error)
+}
+
+type SecretLicenseStore struct {
+	cl             client.Client
+	namespace      string
+	controlPlaneID string
+}
+
+var _ Storer = &SecretLicenseStore{}
+
+//+kubebuilder:rbac:groups="",resources=secrets,verbs=get;update
+
+// NewSecretLicenseStore creates a storage to store the fetched license to a secret.
+func NewSecretLicenseStore(cl client.Client, namespace, controlPlaneID string) *SecretLicenseStore {
+	return &SecretLicenseStore{
+		cl:             cl,
+		namespace:      namespace,
+		controlPlaneID: controlPlaneID,
+	}
+}
+
+// Store stores license to the secret `konnect-license-<cpid>`.
+func (s *SecretLicenseStore) Store(ctx context.Context, l license.KonnectLicense) error {
+	secret := &corev1.Secret{}
+	err := s.cl.Get(ctx, k8stypes.NamespacedName{
+		Namespace: s.namespace,
+		Name:      licenseResourceNamePrefix + s.controlPlaneID,
+	}, secret)
+	if err != nil {
+		return err
+	}
+	// TODO: set labels/annotations of the secret?
```

Comment:
We could use a label like `managed-by: ingress-controller` for this but I'm open to suggestions.

Reply:
I agree to add some labels or annotations but not sure what to add. This is a good suggestion.